### PR TITLE
App Link  Install Support

### PR DESCRIPTION
**Pull request by sojanpr**
Adding support for FB deferred app link Data

@aaustin

### CODE REVIEW

File: Branch-SDK/src/io/branch/referral/ServerRequestRegisterInstall.java [hunk 1]
```diff
@@ -94,6 +95,7 @@ public void onRequestSucceeded(ServerResponse resp, Branch branch) {
             prefHelper_.setExternalIntentExtra(PrefHelper.NO_STRING_VALUE);
             prefHelper_.setAppLink(PrefHelper.NO_STRING_VALUE);
             prefHelper_.setPushIdentifier(PrefHelper.NO_STRING_VALUE);
+            prefHelper_.setIsAppLinkTriggeredInit(false);
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/ServerRequestRegisterInstall.java]: The primary pathway this can be triggered on on install. We'll need to pass the boolean up here.

Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/ServerRequestRegisterInstall.java]: oops - I misread .. sorry. But you'll need this in the `onRequestSucceeded` for RegisterOpen too I believe, right?

File: Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+class DeferredAppLinkDataHandler {
+    private static final String NATIVE_URL_KEY = "com.facebook.platform.APPLINK_NATIVE_URL";
+
+    public static Boolean fetchDeferredAppLinkData(Context context, final AppLinkFetchEvents callback) {
+        boolean isRequestSucceeded = true;
+        try {
+            // Init FB SDK
+            Class<?> FacebookSdkClass = Class.forName("com.facebook.FacebookSdk");
+            Method initSdkMethod = FacebookSdkClass.getMethod("sdkInitialize", Context.class);
+            initSdkMethod.invoke(null, context);
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java]: Did you find that this was required? I think I was able to get it to work without calling init on the FB SDK.

File: Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+            Object completionListenerInterface = Proxy.newProxyInstance(AppLinkDataCompletionHandlerClass.getClassLoader()
+                    , new Class<?>[]{AppLinkDataCompletionHandlerClass}
+                    , ALDataCompletionHandler);
+
+            String fbAppID = context.getString(context.getResources().getIdentifier("facebook_app_id", "string", context.getPackageName()));
+            if (TextUtils.isEmpty(fbAppID)) {
+                isRequestSucceeded = false;
+            } else {
+                fetchDeferredAppLinkDataMethod.invoke(null, context, fbAppID, completionListenerInterface);
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java]: Is it possible to use the method w/o the app id? That was the one I'm using in BMF.

File: Branch-SDK/src/io/branch/referral/Branch.java [hunk 4]
```diff
@@ -2949,9 +2954,43 @@ private void initializeSession(BranchReferralInitListener callback) {
...
+                        if (!TextUtils.isEmpty(bncLinkClickId)) {
+                            prefHelper_.setLinkClickIdentifier(bncLinkClickId);
+                            prefHelper_.setIsAppLinkTriggeredInit(true);
+                        }
+                    }
+                    BranchReferralInitListener deferredCallback = null;
+                    if (deferredInitListener_ != null) {
+                        deferredCallback = deferredInitListener_.get();
+                    }
+                    registerAppInit(deferredCallback);
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Can you test in airplane mode and the like to ensure that the Facebook SDK always calls back. We don't want them to introduce an edge case that causes us to fail to initialize.

File: Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java (hunk 2 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java]: I have seen the error "Unable to fetch deferred params since SDK is not initialized" .I couldn't find the exact error checking the code but FB code seems like calling initialize if we don't do it when calling `fetchDeferredParams`

We wanted to catch all possible error before calling `fetchDeferredParams` to avoid waiting for errors in callback

File: Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java (hunk 3 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java]: The methods with out app id internally calls `fetchDeferredAppLinkFromServer(Context context, String applicationId, AppLinkData.CompletionHandler completionHandler) ` method and they read the app id (The `facebook_app_id` from string resource) .
IN case user  does not have it provided , using this approach we can find it before hands rather than waiting for the callback.

File: Branch-SDK/src/io/branch/referral/ServerRequestRegisterInstall.java (hunk 1 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/ServerRequestRegisterInstall.java]: Good catch. Will update this

File: Branch-SDK/src/io/branch/referral/Branch.java (hunk 4 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/Branch.java]: I have already tested with no internet condition and it seems it called back with in 500 ms.

**Comment by sojanpr** [comment]
Fixed  the issue with `facebook_app_link_checked` state. Review  ready

File: Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java (hunk 3 above)
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/DeferredAppLinkDataHandler.java]: Nice

**Comment by aaustin** [comment]
Looks good ! 👍 Please fix conflicts
